ICP: Improve AES-GCM performance

AttilaFueloep opened the pull request:
### Motivation and Context

Currently SIMD accelerated AES-GCM performance is limited by two factors:

a. The need to disable preemption and interrupts and save the FPU state before using it and to do the reverse when done. Due to the way the code is organized (see (b) below) we have to pay this price twice for each 16 byte GCM block processed.

b. Most processing is done in C, operating on single GCM blocks. The use of SIMD instructions is limited to the AES encryption of the counter block (AES-NI) and the Galois multiplication (PCLMULQDQ). This leads to the FPU not being fully utilized for crypto operations.

### Description

To solve (a) we do crypto processing in larger chunks while owning the FPU. An `icp_gcm_avx_chunk_size` module parameter was introduced to make this chunk size tweakable. It defaults to 32 KiB. This step alone roughly doubles performance. (b) is tackled by porting and using the highly optimized openssl AES-GCM assembler routines, which do all the processing (CTR, AES, GMULT) in a single routine. Both steps together result in up to 32x reduction of the time spend in
the en/decryption routines, leading up to approximately 12x throughput increase for large (128 KiB) blocks.

### How Has This Been Tested?

During development a special version was prepared which ran old an new routines in succession and compared the results. This proved to be quite helpful in finding bugs.

The final version was tested with the ZTS, a six hours run of zloop,  rsyncing an Arch Linux installation (60 GiB) across multiple updates and running scrubs with the old and new code in between. Finally I run this version on the encrypted ZFS root box I'm typing this on. 

### Performance figures

The tables below show the speed improvements compared to the original implementation (pclmulqdq). The data was captured using  [bpftrace](https://github.com/iovisor/bpftrace) scripts which can be found [here](https://gist.github.com/AttilaFueloep/717ce0aef63db64c7e9d13ee93b1b93b) and manually post-processed.

#### Table 1:

Time spend in `gcm_decrypt_final()` and in `gcm_mode_encrypt_contiguous_blocks()` in nano seconds. This basically covers only aes-gcm processing in the decrypt case and processing and write out in the encrypt case. Only some exemplary data sizes are shown.

```
+-------+--------------------+-------+--------------------+-------+
| Size  | Decrypt time [ns]  | Fact. | Encrypt time [ns]  | Fact. |
| [KiB] +-----------+--------+       +-----------+--------+       |
|       | pclmulqdq |  avx   |       | pclmulqdq |  avx   |       |
+-------+-----------+--------+-------+-----------+--------+-------+
|   0.5 |     8,164 |  1,252 |   6.5 |    12,999 |  4,818 |   2.7 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     1 |    16,040 |  1,427 |  11.2 |    24,846 |  5,392 |   4.6 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     2 |    29,727 |  1,710 |  17.4 |    49,764 |  6,214 |   8.0 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     4 |    57,583 |  2,364 |  24.4 |    96,362 |  7,444 |  12.9 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     8 |   137,249 |  3,890 |  35.5 |   186,323 |  9,284 |  20.1 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    16 |   238,991 |  6,028 |  39.6 |   390,489 | 16,117 |  24.2 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    32 |   546,365 | 11,990 |  45.9 |   768,231 | 29,139 |  26.4 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    64 | 1,094,133 | 22,768 |  48.1 | 1,447,619 | 48,879 |  29.6 |
+-------+-----------+--------+-------+-----------+--------+-------+
|   128 | 2,038,599 | 42,521 |  47.9 | 2,797,580 | 83,720 |  33.4 |
+-------+-----------+--------+-------+-----------+--------+-------+
```

For small sizes I could only capture encryption times as follows:

```
+---------+--------------------+-------+
|  Size   | Encrypt time [ns]  | Fact. |
| [Bytes] +-----------+--------+       +
|         | pclmulqdq |  avx   |       |
+---------+-----------+--------+-------+
|     16  |       819 |    900 |   0.9 |
+---------+-----------+--------+-------+
|     32  |     1,074 |  1,091 |   1.0 |
+---------+-----------+--------+-------+
|    128  |     3,015 |    954 |   3.2 |
+---------+-----------+--------+-------+
```

#### Table 2:

Time spend processing a `gcm_cxt` in nano seconds. This starts with the call to `gcm_init_ctx()` and ends with the return from `gcm_{en,de}crypt_final()`, thus covering context initialization, final tag calculation and write out. Again only some exemplary data sizes are shown.

```
+-------+--------------------+-------+--------------------+-------+
| Size  | Decrypt time [ns]  | Fact. | Encrypt time [ns]  | Fact. |
| [KiB] +-----------+--------+       +-----------+--------+       |
|       | pclmulqdq |  avx   |       | pclmulqdq |  avx   |       |
+-------+-----------+--------+-------+-----------+--------+-------+
|   0.5 |    10,869 | 3,066  |   3.6 |    13,809 |  5,451 |   2.5 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     1 |    17,798 |  3,107 |   5.7 |    25,400 |  5,929 |   4.3 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     2 |    34,107 |  3,365 |  10.1 |    48,947 |  6,579 |   7.4 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     4 |    66,349 |  4,233 |  15.7 |    94,665 |  8,235 |  11.5 |
+-------+-----------+--------+-------+-----------+--------+-------+
|     8 |   153,649 |  7,150 |  21.5 |   182,939 | 11,122 |  16.4 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    16 |   287,996 | 11,126 |  25.9 |   373,888 | 16,842 |  22.2 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    32 |   605,206 | 21,532 |  28.1 |   751,444 | 31,245 |  24.1 |
+-------+-----------+--------+-------+-----------+--------+-------+
|    64 | 1,268,601 | 39,598 |  32.0 | 1,401,151 | 51,788 |  27.1 |
+-------+-----------+--------+-------+-----------+--------+-------+
|   128 | 2,037,635 | 73,167 |  27.8 | 2,712,798 | 92,379 |  29.4 |
+-------+-----------+--------+-------+-----------+--------+-------+
```

Here the nuber of captured small blocks were to low to show.

#### Fio results:

The user visible speedup is about 12 times. This was measured with `fio` runs (280 MiB / 22 MiB = 12.7) and manually copying and dd'ing files and measuring time taken.

```
# fio --name="Test ICP PCLMUL" --randrepeat=1 --ioengine=libaio
--gtod_reduce=1 --directory=/var/tmp --bs=512 --iodepth=64 --filesize=10M-100m
--readwrite=randrw --rwmixread=50 --percentage_random=50
--blocksize_range=512-128k --blocksize_unaligned --unlink=1 --nrfiles=250

Test ICP: (g=0): rw=randrw, bs=(R) 512B-128KiB, (W) 512B-128KiB, (T) 512B-128KiB, ioengine=libaio, iodepth=64
fio-3.16
Starting 1 process
Test ICP: Laying out IO files (250 files / total 13552MiB)
Jobs: 1 (f=168): [m(1)][100.0%][r=27.1MiB/s,w=28.4MiB/s][r=484,w=521 IOPS][eta 00m:00s]
Test ICP: (groupid=0, jobs=1): err= 0: pid=453146: Tue Dec 17 19:41:51 2019
  read: IOPS=390, BW=22.2MiB/s (23.3MB/s)(6763MiB/303985msec)
   bw (  KiB/s): min= 9460, max=33184, per=99.90%, avg=22758.74, stdev=3955.12, samples=607
   iops        : min=  142, max=  582, avg=389.86, stdev=75.40, samples=607
  write: IOPS=389, BW=22.3MiB/s (23.4MB/s)(6790MiB/303985msec); 0 zone resets
   bw (  KiB/s): min= 8943, max=31526, per=99.90%, avg=22849.19, stdev=4078.35, samples=607
   iops        : min=  146, max=  550, avg=389.24, stdev=76.27, samples=607
  cpu          : usr=0.25%, sys=88.38%, ctx=22414, majf=0, minf=7
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=118633,118465,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=22.2MiB/s (23.3MB/s), 22.2MiB/s-22.2MiB/s (23.3MB/s-23.3MB/s), io=6763MiB (7091MB), run=303985-303985msec
  WRITE: bw=22.3MiB/s (23.4MB/s), 22.3MiB/s-22.3MiB/s (23.4MB/s-23.4MB/s), io=6790MiB (7120MB), run=303985-303985msec

# fio --name="Test ICP AVX" --randrepeat=1 --ioengine=libaio
--gtod_reduce=1  --directory=/var/tmp --bs=512 --iodepth=64 --filesize=10M-100m
--readwrite=randrw --rwmixread=50 --percentage_random=50
--blocksize_range=512-128k --blocksize_unaligned --unlink=1 --nrfiles=250

Test ICP: (g=0): rw=randrw, bs=(R) 512B-128KiB, (W) 512B-128KiB, (T) 512B-128KiB, ioengine=libaio, iodepth=64
fio-3.16
Starting 1 process
Test ICP: Laying out IO files (250 files / total 13552MiB)
Jobs: 1 (f=169): [m(1)][100.0%][r=219MiB/s,w=225MiB/s][r=4078,w=4147 IOPS][eta 00m:00s]
Test ICP: (groupid=0, jobs=1): err= 0: pid=62831: Tue Dec 17 18:48:44 2019
  read: IOPS=4890, BW=279MiB/s (292MB/s)(6763MiB/24256msec)
   bw (  KiB/s): min=115495, max=420832, per=100.00%, avg=286461.67, stdev=86508.74, samples=48
   iops        : min= 2110, max= 7236, avg=4906.65, stdev=1447.97, samples=48
  write: IOPS=4883, BW=280MiB/s (294MB/s)(6790MiB/24256msec); 0 zone resets
   bw (  KiB/s): min=118692, max=415220, per=100.00%, avg=287503.60, stdev=86732.52, samples=48
   iops        : min= 2090, max= 7242, avg=4896.92, stdev=1444.53, samples=48
  cpu          : usr=3.20%, sys=77.05%, ctx=5272, majf=0, minf=8
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=118633,118465,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=279MiB/s (292MB/s), 279MiB/s-279MiB/s (292MB/s-292MB/s), io=6763MiB (7091MB), run=24256-24256msec
  WRITE: bw=280MiB/s (294MB/s), 280MiB/s-280MiB/s (294MB/s-294MB/s), io=6790MiB (7120MB), run=24256-24256msec
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

AttilaFueloep commented:
Some further remarks:

- Since this is crypto I tried to keep my changes to the existing code at a bare minimum. This should also make it easier to merge possible future upstream changes.

- I left in  a couple of fixmes in places where I would appreciate reviewer feedback.

- The changes done to the openssl assembler routines during porting can be found [here (aesni-gcm-x86_64.S.diff)](https://gist.github.com/AttilaFueloep/c1aa86170adfbf11fdb37a2ffcad8aa3) and [here (ghash-x86_64.S.diff)](https://gist.github.com/AttilaFueloep/278697313a980f64260965c0276c40ad)

- The ICP module parameters doesn't seem to be documented, so I have no place to add documentation to.

- My testing resources are quite limited so I would really appreciate any further testing.

behlendorf commented:
Very impressive!  This is a huge performance improvement, thank you for tackling this work.  

cc: @jasonbking @Fabian-Gruenbichler

AttilaFueloep commented:
I'm really sorry for the build failures, but I can't reproduce them locally. Do I need to add anything besides `--enable-debug --enable-debuginfo` to configure?

behlendorf commented:
No problem, that's why we have to CI to test across a wide range of kernel, compiler, and library versions.  It looks like you've managed to sort it out.  Using `--enable-debug --enable-debuginfo` should be sufficient.

AttilaFueloep commented:
Yes, forgot to run `make cstyle` after my last change and my editor converted four to two spaces. `make lint` gives me tons of errors even in files I did not touch, seems Arch Linux `cppcheck` is special. But I think I can fix the errors.

behlendorf commented:
> make lint gives me tons of errors even in files I did not touch

Try rebasing the PR on the master branch from yesterday.  A collection of cppcheck patches were merged yesterday for issues only detected by newer versions of cppcheck, up to cppcheck v1.88 should now be clean.

AttilaFueloep commented:
Well the PR is based on master@a3640486fffc which is from today but Arch has `cppcheck 1.89-1`. I'll try downgrading it to 1.88 and see if that helps.

behlendorf commented:
The CI is still using cppcheck 1.82 and is reporting [these](http://build.zfsonlinux.org/builders/Ubuntu%2017.10%20x86_64%20%28STYLE%29/builds/21526/steps/shell_3/logs/stdio) warnings which do look related to this change.

AttilaFueloep commented:
Right, although I didn't fully understand the error, I tried to fix them with commit f488b0e which didn't help. I now downgraded  cppcheck to 1.86 and it's silent with and without f488b0e. May that be a false positive? Paxcheck complained about an executable stack in the assembler files, which I've already fixed but not pushed yet. Should I revert f488b0e and add a /* LINTED */ comment?

AttilaFueloep commented:
Well, I see ztest failing in the new code. I've to have a look tomorrow.

behlendorf commented:
That's strange, I wasn't able to reproduce the issue with cppcheck 1.88 either.  Let's revert f488b0e from the PR for now, we'll be updating the version of cppcheck used pretty shortly so this may not be an issue.

AttilaFueloep commented:
Ok, will do tomorrow. Looks like there is no support for the `MOVBE` instruction in the testers. I'm a bit puzzled here. What virtual CPU do they use?

behlendorf commented:
@AttilaFueloep nothing exotic, they're currently using ec2 m3.large instances and will soon to be switched to m5.large instances.

AttilaFueloep commented:
Ok, thanks. I've to do some research regarding `MOVBE` support.

AttilaFueloep commented:
Ok, M3 instances use Ivy/Sandy Bridge CPUs which do have AVX but no MOVBE. I'd need to update the requirements from AVX to AVX2 (or test explicitly for MOVBE), which would mean we can't use the testers to test this PR. This would lift CPU requirements to at least Haswell or Excavator. I've no idea how to proceed from here.

behlendorf commented:
Actually, it sounds like this went pretty well since we catch this MOVBE issue.  What you're going to want to do is a check for `X86_FEATURE_MOVBE` in `include/os/linux/kernel/linux/simd_x86.h`.  This way we'll be able to detect if it's available at run time, take a look at the `zfs_*_available()` functions.  Are there other specific instructions we should be concerned about?

As for how to test this I don't think it'll be long before switching to m5 instances which according to the documentation use skylake.

AttilaFueloep commented:
Luckily, yes. I've to go over the openssl requirements again, obviously I missed something there.

This also poses the question if this code should always be compiled in, regardless of the build environment detected by `configure` and only called on supported platforms at run time. The pclmulqdq code gets compiled in only if building on supported hardware (`defined(HAVE_PCLMULQDQ)`), I just mirrored that behaviour.

behlendorf commented:
> should always be compiled in, regardless of the build environment detected by configure and only called on supported platforms at run time.

Yup.  And it looks like that is the case for pclmulqdq  code so your code is pretty close.  There are a couple level of checks going on here which can be a bit confusing.  But to summarize with an example, the `HAVE_PCLMULQDQ` check detects if the host build tool chain understands instruction, not if the host hardware supports it.  This is coupled with `zfs_pclmulqdq_available()` which is used for the run time check.

You'll want to add similar checks for `MOVBE` to `config/toolchain-simd.m4` and `include/os/linux/kernel/linux/simd_x86.h` and `lib/libspl/include/sys/simd.h`.  Then always compile the code in when `HAVE_MOVBE` but only enable it on module load when `zfs_pclmulqdq_available()` is true.

Fabian-Gruenbichler commented:
I probably won't have time until the new year to take a closer look, but from a quick glance it seems like the proper preconditions are in place to avoid long kfpu_begin/end blocks which is probably all I can really contribute here ;)

AttilaFueloep commented:
@behlendorf All right, got it. Checking for tool chain support definitely makes more sense. I'll try to find some time tomorrow.

AttilaFueloep commented:
The "signed integer overflow" lint error is a false positive and will go away with cppcheck >= 0.86. Since the testers do not run the new code now, (they are missing the `MOVBE` instruction), I don't think the test failure is related to my changes. All other tests pass, but they are not very meaningful until the testers will be updated to newer ec2 instances which will be able to test the added code.

behlendorf commented:
@AttilaFueloep thanks for addressing the review feedback so quickly.  I agree, the testing failures you hit were unrelated to this change.

The CI has been updated to use m5d.large instances which should support the needed instruction.  I've gone ahead and resubmitted this PR for a fresh test run, but you may want to rebase it as well to clean out any old results.  Additionally, the cppcheck version has been updated that issue should be resolved.

AttilaFueloep commented:
@behlendorf 

> The CI has been updated to use m5d.large instance

Fantastic!

Squashed and rebased, it should be ready for review now.

rlaager commented:
@AttilaFueloep Any chance you have a comparison between AES-CCM and AES-GCM in ZFS? (It's not strictly related to this PR, but you may be in a good position to easily test it, which would be helpful to me in a different context.)

AttilaFueloep commented:
@rlaager Generally speaking I think the answer to this stack exchange [question](https://crypto.stackexchange.com/questions/6842/how-to-choose-between-aes-ccm-and-aes-gcm-for-storage-volume-encryption) sums it up pretty well. (Please note that the mentioned  parallelization isn't implemented in the ICP.) If I had a choice I would always go for AES-GCM because of its encrypt-then-MAC scheme, which reduces attack surface. The major weak point may be the possibility of compromising the hash subkey which will completely destroy authentication assurance, opening the window for counter mode attacks. (See the [NIST recommendations, appendices A and B](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf).) But AFAIK attacks have only been shown for shortened  authentication tags which ZFS does not use, and since AES-GCM is pretty much standard these days I feel sufficiently safe.

Performance wise, one AES and one GHASH operation is usually faster than two AES operations, at least on CPUs which accelerate in hardware both, AES and a carry-less quadword multiplication. I did run the `fio` tests above on a aes-256-gcm and a aes-256-ccm dataset and the GCM run is approximately 1.15 times faster than the CCM run. Please also note that this PR doesn't improve AES-CCM performance, so if this gets merged, the speed difference will be much larger.

```
# fio --name="Test ICP CCM" --randrepeat=1 --ioengine=libaio --gtod_reduce=1 \
  --directory=/test-ccm --bs=512 --iodepth=64 --filesize=10M-100m --readwrite=randrw \
  --rwmixread=50 --percentage_random=50 --blocksize_range=512-128k \
  --blocksize_unaligned --unlink=1 --nrfiles=250
Test ICP CCM: (g=0): rw=randrw, bs=(R) 512B-128KiB, (W) 512B-128KiB, (T) 512B-128KiB, ioengine=libaio, iodepth=64
fio-3.16
Starting 1 process
Test ICP CCM: Laying out IO files (250 files / total 13552MiB)
Jobs: 1 (f=168): [m(1)][100.0%][r=24.1MiB/s,w=27.0MiB/s][r=419,w=485 IOPS][eta 00m:00s]
Test ICP CCM: (groupid=0, jobs=1): err= 0: pid=313725: Tue Dec 24 19:57:30 2019
  read: IOPS=373, BW=21.3MiB/s (22.3MB/s)(6763MiB/317285msec)
   bw (  KiB/s): min= 6845, max=31631, per=99.97%, avg=21819.90, stdev=4207.95, samples=634
   iops        : min=  102, max=  578, avg=373.77, stdev=82.53, samples=634
  write: IOPS=373, BW=21.4MiB/s (22.4MB/s)(6790MiB/317285msec); 0 zone resets
   bw (  KiB/s): min= 6522, max=31427, per=99.97%, avg=21906.61, stdev=4162.38, samples=634
   iops        : min=  112, max=  552, avg=373.19, stdev=81.63, samples=634
  cpu          : usr=0.22%, sys=98.75%, ctx=8778, majf=0, minf=7
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=118633,118465,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=21.3MiB/s (22.3MB/s), 21.3MiB/s-21.3MiB/s (22.3MB/s-22.3MB/s), io=6763MiB (7091MB), run=317285-317285msec
  WRITE: bw=21.4MiB/s (22.4MB/s), 21.4MiB/s-21.4MiB/s (22.4MB/s-22.4MB/s), io=6790MiB (7120MB), run=317285-317285msec

# fio --name="Test ICP GCM" --randrepeat=1 --ioengine=libaio --gtod_reduce=1 \
  --directory=/var/tmp --bs=512 --iodepth=64 --filesize=10M-100m --readwrite=randrw \
  --rwmixread=50 --percentage_random=50 --blocksize_range=512-128k \
  --blocksize_unaligned --unlink=1 --nrfiles=250
Test ICP GCM: (g=0): rw=randrw, bs=(R) 512B-128KiB, (W) 512B-128KiB, (T) 512B-128KiB, ioengine=libaio, iodepth=64
fio-3.16
Starting 1 process
Test ICP GCM: Laying out IO files (250 files / total 13552MiB)
Jobs: 1 (f=168): [m(1)][100.0%][r=31.3MiB/s,w=33.2MiB/s][r=572,w=604 IOPS][eta 00m:00s]
Test ICP GCM: (groupid=0, jobs=1): err= 0: pid=379524: Tue Dec 24 20:22:02 2019
  read: IOPS=438, BW=24.0MiB/s (26.2MB/s)(6763MiB/270561msec)
   bw (  KiB/s): min= 9869, max=36209, per=99.97%, avg=25586.19, stdev=5142.53, samples=541
   iops        : min=  148, max=  666, avg=438.30, stdev=99.30, samples=541
  write: IOPS=437, BW=25.1MiB/s (26.3MB/s)(6790MiB/270561msec); 0 zone resets
   bw (  KiB/s): min= 8943, max=36518, per=99.96%, avg=25686.92, stdev=5190.41, samples=541
   iops        : min=  146, max=  658, avg=437.63, stdev=99.36, samples=541
  cpu          : usr=0.24%, sys=98.36%, ctx=9416, majf=0, minf=7
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=118633,118465,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=24.0MiB/s (26.2MB/s), 24.0MiB/s-24.0MiB/s (26.2MB/s-26.2MB/s), io=6763MiB (7091MB), run=270561-270561msec
  WRITE: bw=25.1MiB/s (26.3MB/s), 25.1MiB/s-25.1MiB/s (26.3MB/s-26.3MB/s), io=6790MiB (7120MB), 
```

rlaager commented:
@behlendorf If this patch lands, you may want to consider changing `encryption=on` to use GCM instead of CCM, given the large performance gain and possible security benefits.

AttilaFueloep commented:
@rlaager Tom Caputi, the Author of ZFS native encryption, explains the reasoning behind choosing CCM as default mode [here](https://github.com/zfsonlinux/zfs/issues/6707#issuecomment-333683608)

> The attacks you mentioned are very new (as far as I'm aware anyway) and only apply to applications which use a truncated MAC (less than 128 bits). We use the full 128 bits so this is not an issue here. This could represent a weakness in the future, but for right now nobody is aware of any security problems with GCM using 128 bit MACs. This is also the reason we chose CCM as the default mode, simply as a precaution and despite the better performance that GCM offers.

If there's enough interest I could be beaten to port the openssl CCM assembler routines too, but the improvements won't be as big as in the GCM case. Here is the output of `openssl speed` indicating that GCM performs 3-4 times faster then CCM.

```
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-256-gcm     547314.06k  1453827.45k  2611890.86k  3844141.21k  4515258.37k  4579874.13k
aes-256-ccm     190056.97k   535891.22k   883048.54k  1044096.68k  1104838.66k  1109038.42k
-------------------------------------------------------------------------------------------
Factor                 2.9          2.7          2.9          3.7          4.1          4.1
```
So I'd expect a three times speedup for ZFS as opposed to the twelve times seen for GCM (and it's a lot of work to do).

AttilaFueloep commented:
Rebasing to get rid of the test failures since there were commits fixing them recently.

AttilaFueloep commented:
Looks like the codecov runner lacks support for some (or all) of the AVX, MOVBE, AES-NI and PCLMULQDQ instructions, so the new code isn't covered. The Ubuntu 16.04 failures seem unrelated.

behlendorf commented:
That's right.  Unfortunately, the coverage builder has some additional complications which prevented us from updating it along with the other test instances.  Though we intend to eventually to do so.

AttilaFueloep commented:
Thanks for your confirmation Brian. I'll try to generate a code coverage report locally though it may take a while.

adamdmoss commented:
I'm giving this a spin but not seeing speedups so I presume the path isn't being hit for me.  I have avx + pclmulqdq + aes flags in my cpuinfo, though not MOVBE.  Am I doomed to poor perf?

AttilaFueloep commented:
@adamdmoss  Unfortunately, yes. The MOVBE instruction is used in aesni-gcm-x86_64.S and according to the comment in this file "it reduces shuffles pressure with notable relative improvement". Since Haswell and newer architectures support MOVBE I think it's reasonable to use it. At least it's not worth the effort replacing it with move and shuffle instructions, reducing performance on newer CPUs.

adamdmoss commented:
If anyone's interested, I did a local replacement of the MOVBE uses with BSWAP+MOV.
Unable to say how the speed compares with the MOVBE version but my encrypted gcm-256 copies run 3x faster than the trunk implementation and are no longer CPU-limited, so I'm happy enough.  I can probably turn that into an option in the future if there's interest.

adamdmoss commented:
Modulo the local changes I needed, consider me a happy tester/user of this branch. :)

rlaager commented:
Any idea how many CPU families would see an improvement with that (having another option that does not use MOVEBE)?

AttilaFueloep commented:
All CPUs that support [AES-NI](https://en.wikipedia.org/wiki/AES_instruction_set
), [PCLMUL](https://en.wikipedia.org/wiki/CLMUL_instruction_set) and [AVX](https://de.wikipedia.org/wiki/Advanced_Vector_Extensions) but not [MOVBE](https://stackoverflow.com/questions/5246146/which-cpus-support-movbe-instruction). 

CPUs feature support is quite a mess (particularly Intel adds and removes features wildly) but on the Intel side that would be basically Sandy- and Ivy-Bridge (Core i5/i7 2XXX/3XXX and the Xeons based on that architecture) and on the AMD side Bulldozer, Piledriver, and Steamroller.

rlaager commented:
> All CPUs that support [AES-NI](https://en.wikipedia.org/wiki/AES_instruction_set), [PCLMUL](https://en.wikipedia.org/wiki/CLMUL_instruction_set) and [AVX](https://de.wikipedia.org/wiki/Advanced_Vector_Extensions) but not [MOVBE](https://stackoverflow.com/questions/5246146/which-cpus-support-movbe-instruction).
> 
> CPUs feature support is quite a mess (particularly Intel adds and removes features wildly) but on the Intel side that would be basically Sandy- and Ivy-Bridge (Core i5/i7 2XXX/3XXX and the Xeons based on that architecture) and on the AMD side Bulldozer, Piledriver, and Steamroller.

That seems like quite a few. So ideally, it'd be nice to add that version as another alternative, but that doesn't have to hold up this PR.

behlendorf commented on module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:
```diff
@@ -0,0 +1,892 @@
+# project. The module is, however, dual licensed under OpenSSL and
+# CRYPTOGAMS licenses depending on where you obtain it. For further
+# details see http://www.openssl.org/~appro/cryptogams/.
+# ====================================================================
```
Please add the gcm and ghash implementations to the top level `COPYRIGHT` file along with the other exceptions.  You'll want to also add a `THIRDPARTYLICENSE` file for the correct license a well.

AttilaFueloep replied:
Will do.

AttilaFueloep replied:
Done, a337b63.

rlaager commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +972,567 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+ * FIXME: Do we need this? There was a bug were FPU regs leaked to
+ * userspace and they hold sensitive internal crypto data. The performance
+ * impact is measurable. TODO
+ */
```
What about adding a check for the kernel version and not clearing them if the kernel is new enough to be patched, where "new enough" is defined by the mainline version? That'd be the most conservative version check, so you would still clear on some kernels (e.g. LTS kernels) where it's not necessary, but that'd be the safest. And then eventually, the clearing and thus performance impact goes away as people end up on safe kernels. That said, how bad is the performance impact?

AttilaFueloep replied:
I tend to agree with @behlendorf on this. It also gives the added benefit to guard against potential future bugs. Let's see what other reviewers think. If there is consensus I will look if there is a need to clear some registers in the non-avx case as well.

@rlaager The performance impact is a function of payload size and CPU architecture and is hard to measure reliably. Roughly estimating from my measurements (Coffee Lake) it's about 2% in the worst case. For an average workload it should be negligible.

behlendorf replied:
Hearing no objection, let always clear the registers to be absolutely sure.

AttilaFueloep replied:
OK, adjusted the comment in 384bcaf.

AttilaFueloep commented:
@adamdmoss Thanks for trying this out and glad to hear that it works for you. What are your ideas to make your changes optional?

AttilaFueloep commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +972,567 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+}
+
+/*
+ * FIXME: Doesn't seem to be done on the aes side, do we need to do this?
```
As above, if there is consensus, it should be done for the non-avx case as well.

AttilaFueloep commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +972,567 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+
+	/* Allocate a buffer to encrypt to if there is enough input. */
+	if (bleft >= GCM_AVX_MIN_ENCRYPT_BYTES) {
+		ct_buf = vmem_alloc(chunk_size, ctx->gcm_kmflag);
```
Will do.

behlendorf replied:
This isn't critical so if it's not easy to test I don't think we should let this hold anything up.  I simply wanted to mention it.

AttilaFueloep replied:
It's easy to test but just a bit time consuming to do. I like the idea though and think it's worth a try. I'll try to find some time in the near future.

AttilaFueloep commented on module/icp/algs/modes/gcm.c:
```diff
@@ -737,6 +834,12 @@ gcm_impl_init(void)
 	strcpy(gcm_fastest_impl.name, "fastest");
 
+#ifdef CAN_USE_GCM_ASM
+	/* FIXME: Should we make `avx` the `fastest` implementation? */
```
All right, will remove the comment later.

AttilaFueloep replied:
Adjusted comment and a minor tweak. 45ecbc5

adamdmoss commented:
> What are your ideas to make your changes optional?

I was going to do the least redundant thing I could think of - hide the MOVBE in the .S inside a macro and include the same .S twice - once for each variant.  The rest is boilerplate I'd crib from the existing branch changes.

I also agree entirely that there's no reason the proposed variant should hold up getting this branch landed.

AttilaFueloep commented:
Yes that would work as a compile time option. To make this a run time option one would need to copy the entire routine using MOVBE, make the changes there and call it via a function pointer which is initialized according to the CPU capabilties we run on.

Since there is agreement I'd say let's discuss this in a follow up PR.

AttilaFueloep commented on module/icp/algs/modes/gcm.c:
```diff
@@ -30,12 +30,46 @@
+#endif
+#define	GCM_IMPL_READ(i) (*(volatile uint32_t *) &(i))
+static uint32_t icp_gcm_impl = IMPL_FASTEST;
+static uint32_t user_sel_impl = IMPL_FASTEST;
```
I think I know what's going on. Need to verify an find a fix.

AttilaFueloep replied:
Commit 0366890 should fix this issue, at least in my testing  I wasn't able to reproduce with it applied.

AttilaFueloep commented on module/icp/algs/modes/gcm.c:
```diff
@@ -112,6 +152,7 @@ gcm_mode_encrypt_contiguous_blocks(gcm_ctx_t *ctx, char *data, size_t length,
 		ctx->gcm_processed_data_len += block_size;
 
 		if (out == NULL) {
+			/* FIXME: See issue #9661 */
```
Right, will replace the offending code with an assert.

AttilaFueloep replied:
Done, ba7a6b6.

AttilaFueloep commented:
Sorry, just realized that I pushed the MOVBE commit from a incomplete work tree . b233da2 adds the missing bits.

AttilaFueloep commented:
Rebased. This time I skipped the squashing step to make it more clear how the review feedback was addressed. The rebase should also make the Ubuntu 16.04 test failure go away, which popped up randomly in the past.

AttilaFueloep commented:
Hope to have addressed all review feedback besides the vmem vs. kmem alloc one, which is on my todo list. I'd  love to get feedback from Tom on clearing some of the context members.

tcaputi commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +996,562 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+
+	/* Do the bulk encryption in chunk_size blocks. */
+	for (; bleft >= chunk_size; bleft -= chunk_size) {
+		kfpu_begin();
```
I see a few places in the patch where I think we can probably get better performance by doing the `kfpu_begin()` and `kfpu_end()` fewer times. I can't find the commit right now (it must have been from about 2 years ago and it may have just ended up in the main native encryption patch), but when originally testing out the ICP, I found that it was using a ton of cpu just initializing and uninitializing the fpu registers.

Ideally, we would want to do it once per "encryption job" but I suspect we will probably need to do it a little more often due to the way the code is laid out with the `_init()` and `_final()` functions.

You should be able to see this cpu usage by running `perf top` or creating a frame graph while writing out a large sequential file to an encrypted dataset.

AttilaFueloep replied:
Actually this is part of what this patch does. The original code called `kfpu_begin()` and `kfpu_end()` twice per 16 byte GCM block. This patch calls them once per `icp_gcm_avx_chunk_size` which is a module parameter defaulting to 32K (2048 GCM blocks) and raiseable up to 128K (8192 GCM blocks). Since `kfpu_begin()` disables preemption and interrupts we have limit the time we run with them disabled. I'm sure @Fabian-Gruenbichler  can comment on an sane upper limit, maybe an upper limit greater than 128k is acceptable.

AttilaFueloep commented:
@tcaputi 

No worries, thanks for reviewing.

> As part of this change, I believe we should change the default encryption mode to GCM, which should just be a matter of changing `ZIO_CRYPT_ON_VALUE`.

Since there seems to be agreement on that, I'll have a look how to change the default encryption mode.

> Unfortunately, one thing I don't really feel comfortable doing is reviewing the assembly, which seems to be the majority of the code.

Well, it's the openssl assembly code with just a few minor adaptions to make it work with the ICP API. Functionally it should behave exactly as the openssl one does. You can look at the comments in the [aesni-gcm-x86_64.S](https://gist.github.com/AttilaFueloep/c1aa86170adfbf11fdb37a2ffcad8aa3) and [ghash-x86_64.S](https://gist.github.com/AttilaFueloep/278697313a980f64260965c0276c40ad) diffs to get an idea what was changed.

Having said that, I really would appreciate a review of the openssl assembly changes I  did.

AttilaFueloep commented:
If @dot-asm, the author of the marvellous openssl AES-GCM assembly, would be so kind to have a look at the above mentioned diffs, that would be really great.

Fabian-Gruenbichler commented:
On January 22, 2020 1:33 am, Attila Fülöp wrote:
> AttilaFueloep commented on this pull request.
> 
> 
> 
>> +		 * See gcm_mode_encrypt_contiguous_blocks() above and
> +		 * https://github.com/zfsonlinux/zfs/issues/9661
> +		 */
> +		ASSERT(out != NULL);
> +		rv = crypto_put_output_data(tmp, out, block_size);
> +		out->cd_offset += block_size;
> +		gcm_incr_counter_block(ctx);
> +		ctx->gcm_processed_data_len += block_size;
> +		bleft -= need;
> +		datap += need;
> +		ctx->gcm_remainder_len = 0;
> +	}
> +
> +	/* Do the bulk encryption in chunk_size blocks. */
> +	for (; bleft >= chunk_size; bleft -= chunk_size) {
> +		kfpu_begin();
> 
> Actually this is part of what this patch does. The original code called `kfpu_begin()` and `kfpu_end()` twice per 16 byte GCM block. This patch calls them once per `icp_gcm_avx_chunk_size` which is a module parameter defaulting to 32K (2048 GCM blocks) and raiseable up to 128K (8192 GCM blocks). Since `kfpu_begin()` disables preemption and interrupts we have limit the time we run with them disabled. I'm sure @Fabian-Gruenbichler  can comment on an sane upper limit, maybe an upper limit greater than 128k is acceptable. 

I guess as long as it's just an upper limit, it's just one more knob 
that serves as a foot gun ;) AFAICT, this could completely stall the 
system if it gets scheduled on many/most/all cores at the same time, and the 
chunk size is too big. since the new default is already a factor 2048 up 
from the old version, I'd probably wait with bumping the limit until 
there's more concrete evidence that it gives a non-neglible performance 
improvement.

AttilaFueloep commented:
Agreed, especially since the returns are diminishing with larger chunk sizes.

behlendorf commented:
> Having said that, I really would appreciate a review of the openssl assembly changes I did.

I'm no expert either but [these assembly changes](https://gist.github.com/AttilaFueloep/c1aa86170adfbf11fdb37a2ffcad8aa3) makes sense to me given the comments which were very helpful.  Plus they're relatively modest which is great.  Perhaps we can ask @rdolbeau to have a look.  He's been kind enough to contribute much of the optimized raidz and fletcher code.

rdolbeau commented:
@behlendorf I'm more comfortable with higher-level implementations than raw assembly, and I'm not aware of the algorithmic changes from 'normal' AES-GCM to whatever ICP uses... That being said, the changes make sense to me w.r.t. the comments in the code, as well.

From various comments above - 'movbe' should be tested on its own it's not part of AVX2 (as mentioned, some CPUs have MOVBE but not AVX2). And there should be support for non-MOVBE system if possible, as it's a nice boost but not a critical aspect of performance in this patch and such systems are still widely used.

AttilaFueloep commented:
>  'movbe' should be tested on its own

Yes, I already added that.

> And there should be support for non-MOVBE system

That's planned as a follow up PR.

jasonbking commented on module/icp/include/modes/modes.h:
```diff
@@ -189,26 +199,41 @@ typedef struct ccm_ctx {
 	uint8_t *gcm_pt_buf;
 	int gcm_kmflag;
+#ifdef CAN_USE_GCM_ASM
+	boolean_t gcm_use_avx;
```
You could use a but in common_ctx->cc_flags if you wanted.

AttilaFueloep replied:
That's a nice idea, it didn't occur to me while writing this. Let me think about it.

AttilaFueloep replied:
I looked at my notes and realized I decided against that, mainly for two reasons, first it adds a few additional cycles and second, strictly speaking, it's not a shared option among all modes, while the struct is already so big that additional four bytes shouldn't matter much.

jasonbking commented:
@skiselkov did a lot of work on this stuff a long time ago that was a bit more extensive.. I've been trying (unsuccessfully) to reach him about taking a look at the work he did (and possibly resurrect/merge it in) since it seems to overlap a decent amount with this (though covers more than GCM, and from his old blog post sounds like it improves the API a bit).. if anyone knows him you might see if he'd be interested in taking a look..

AttilaFueloep commented:
Yes, I found his work and looked at it ([illumos issue](https://www.illumos.org/issues/4896),  [blog](http://zfs-create.blogspot.com/2014/05/optimizing-illumos-kernel-crypto.htm) and [patch](https://github.com/skiselkov/illumos-gate/tree/4896_aes_fast)). What he basically did (IIRC), was to increase the amount of data processed while "owning the FPU", resembling "chunk size" in this PR. This step alone roughly doubles performance.

The major performance gain in this PR is from using the openssl assembly which adds another 6x, resulting in a total of 12x. I decided to skip CCM mode since it's openssl assembly implementation is 4 times slower than the GCM one, so if your going for performance you'll use GCM anyhow. (Just a cost/benefit ratio issue, I must admit.)

AttilaFueloep commented:
Changed the default encryption mode to GCM, as requested. 

@tcaputi  Thanks for the pointer to `ZIO_CRYPT_ON_VALUE`, easiest patch ever, just had to change a single character ;-)

rlaager commented:
Presumably we need to update the man page too. Are there any tests that need updating?

AttilaFueloep commented:
Doh, of course, thanks for reminding me. Not sure about `functional/rsend/send_encrypted_props.ksh` and `functional/cli_root/zpool_split/zpool_split_encryption.ksh`, they use aes-256-ccm explicitly but should work as they are. Should I change them too?

AttilaFueloep commented:
By the way, can I abort a test runner, since I know it will fail anyhow?

behlendorf commented:
> By the way, can I abort a test runner, since I know it will fail anyhow?

Unfortunately not today. 

As for updating those ZTS tests why don't we leave them as is to help ensure we have test coverage for the non-default option.

I'll see about performing some additional testing with this patch over the weekend.

AttilaFueloep commented:
> As for updating those ZTS tests why don't we leave them as is to help ensure we have test coverage for the non-default option.

Agreed.


> I'll see about performing some additional testing with this patch over the weekend.

Take your time.

AttilaFueloep commented:
Sorry, I'm currently not up to my standards. The latest commit should fix the failing test.

behlendorf commented:
The following build warnings are being reported by `objtool` in the new optimized code.  They may be harmless, but they'll need to be investigated and resolved.

```
.../zfs/module/icp/asm-x86_64/modes/aesni-gcm-x86_64.o: warning: objtool: .text+0xd5d: can't find jump dest instruction at .text+0xdcc
.../zfs/module/icp/asm-x86_64/modes/ghash-x86_64.o: warning: objtool: .text+0xa8d: can't find jump dest instruction at .text+0xa5d
```

AttilaFueloep commented:
I already investigated them while developing. They are caused by the constants which are defined in the text section of the assembly file using `.byte` instructions (e.g. `bswap_mask`). The assembler tries to interpret them as opcodes and obviously fails doing so.

AttilaFueloep commented:
Not sure if the test failures are caused by changing the default encryption mode to GCM. Centos 7 fails in the `pool_checkpoint/checkpoint_discard_busy` zfstest, that seems unrelated. Fedora 30 and the Coverage tester have no gdb output but an `zdb: can't open 'ztest': No such file or directory` error, not sure what to make out of this.

adamdmoss commented:
Minor data point - I've been running the HEAD of this branch (albeit with tweaks to make it not require MOVBE) since Jan 9 with good success.  A mere desktop system, but with about 30TB of transfers between pools/datasets with aes-256-gcm in that time.  (Data spring cleaning! :) )

behlendorf commented:
@AttilaFueloep I see, thanks for investigating.  In that case, you can silence the warnings by including the following in the `Makefile.in`:

```diff
diff --git a/module/icp/Makefile.in b/module/icp/Makefile.in
index 123d795..51e69bd 100644
--- a/module/icp/Makefile.in
+++ b/module/icp/Makefile.in
@@ -61,6 +61,13 @@ $(MODULE)-$(CONFIG_X86) += algs/modes/gcm_pclmulqdq.o
 $(MODULE)-$(CONFIG_X86) += algs/aes/aes_impl_aesni.o
 $(MODULE)-$(CONFIG_X86) += algs/aes/aes_impl_x86-64.o
 
+# Suppress objtool "can't find jump dest instruction at" warnings.  They
+# are caused by the constants which are defined in the text section of the
+# assembly file using .byte instructions (e.g. bswap_mask).  The objtool
+# utility tries to interpret them as opcodes and obviously fails doing so.
+OBJECT_FILES_NON_STANDARD_aesni-gcm-x86_64.o := y
+OBJECT_FILES_NON_STANDARD_ghash-x86_64.o := y
+
 ICP_DIRS = \
        api \
        core \
```

I agree, the last round of test failures were unrelated to these changes.  But for good measure I've resubmitted those builds just confirm they aren't consistently reproducible.

This PR has also passed all the testing I've been able to throw at it.

AttilaFueloep commented:
Thanks Brian, I applied your patch and verified that the `objtool` warning are gone now. I second your call for additional reviews.

behlendorf commented:
@adamdmoss thanks for the feedback, that's great to hear.  I haven't observed any issues in my local testing either, this is holding up well.

@AttilaFueloep  what I'd like to do is propose that we go ahead and merge this change on Feb 10th unless an issue is uncovered by further testing.  That gives potential reviewer another full week to comment.

AttilaFueloep commented:
OK, I've gone ahead and squashed and rebased on master. Maybe @tcaputi and @jasonbking would like to have a second look.

Once this is merged I'm going to prepare two follow ups. One for supporting architectures without MOVBE and a second one for clearing more sensitive data in the non-avx cases. I'll also look into documenting the existing and newly added ICP module parameters.

rlaager commented on module/icp/include/modes/modes.h:
```diff
@@ -189,26 +199,41 @@ typedef struct ccm_ctx {
 	size_t gcm_pt_buf_len;
 	uint32_t gcm_tmp[4];
+	/*
+	 * The relative positions of gcm_ghash,, gcm_H and pre-computed
```
nit: extra comma

AttilaFueloep replied:
Done

rlaager commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +996,562 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+
+/*
+ * Finalize the encryption: Zero fill, encrypt, hash and write out an eventual
+ * incomplete last block. Encrypt the ICB. calculate the tag and write it out.
```
nit: capitalize "Calculate"

AttilaFueloep replied:
Done

rlaager commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +996,562 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+	}
+
+	kfpu_begin();
+	/* Pad last incomplete block with zeros,encrypt and hash. */
```
nit: add space after comma

AttilaFueloep replied:
Done

rlaager commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +996,562 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+		return (rv);
+
+	out->cd_offset += ctx->gcm_tag_len;
+	/* FIXME: Doesn't seem to be done on the AES side. */
```
Is this one of the things to do in a follow-up?

AttilaFueloep replied:
Thanks for catching that. Yes I'll add context clearing for the non-avx cases as a follow-up. Rephrased the comments.

rlaager commented on module/icp/algs/modes/gcm.c:
```diff
@@ -845,4 +996,562 @@ icp_gcm_impl_get(char *buffer, zfs_kernel_param_t *kp)
+	clear_fpu_regs();
+	kfpu_end();
+
+	/* compare the input authentication tag with what we calculated */
```
nit: capitalize "Compare"

AttilaFueloep replied:
Done

behlendorf commented:
Since everything has held up well in testing, I've gone ahead and merged this PR as planned.

AttilaFueloep commented:
@behlendorf Thanks for mentioning the changed  default encryption algorithm in the commit message.

split-n commented:
Is there any plans to release a point release with this patch, rather than OpenZFS 2.0 that may contain many other changes? Current encrypt/decrypt speed(read under 400MB/s by EPYC 7302) prevents me from creating an encrypted fs, even after 0.8.3 SIMD support.

nickcmaynard commented:
> Is there any plans to release a point release with this patch, rather than OpenZFS 2.0 that may contain many other changes? Current encrypt/decrypt speed(read under 400MB/s by EPYC 7302) prevents me from creating an encrypted fs, even after 0.8.3 SIMD support.

I would be very glad to see a 0.8.x release with this within it.  The recent kernel and SIMD changes seem to have significantly impacted encryption performance.  This would go some way to fixing the issue.

adamdmoss commented:
Can this be closed?  Or is it open for a reason?

behlendorf commented:
This issue was closed when the PR was merged.  @nickcmaynard we'll consider it for the next 0.8. release.

nickcmaynard commented:
Thank you @behlendorf, much appreciated!
